Rollback time resource during rollback

nick-e opened the pull request:
Ensure that the `Fixed<Time>` resource is accurate during a rollback by rollbacking the time resource when a rollback starts and advancing the time after each rollback tick. This is a continuation of https://github.com/cBournhonesque/lightyear/pull/619. This PR takes into account `Time<Fixed>`'s overstep.

I originally thought that using `add_resource_rollback::<Time<Fixed>>()` and running the `RunFixedMainLoop` schedule would be the solution but `RunFixedMainLoop` requires that the real time clock also progress in order to correctly advance the `Time<Fixed>` resource.

Fixes https://github.com/cBournhonesque/lightyear/issues/618

nick-e commented:
I tested the fix using the replication process found in https://github.com/cBournhonesque/lightyear/issues/618

cBournhonesque commented on lightyear/src/protocol/component.rs:
```diff
@@ -1108,10 +1107,10 @@ impl AppComponentExt for App {
         }
     }
 
+    /// Do not use `Time<Fixed>` for `R`. `Time<Fixed>` is already rollbacked.
```
This comment probably shouldn't be there, right? This is for users who want to use `add_resource_rollback`

nick-e replied:
Updated the doc comments to add clarity. Let me know what you think!

cBournhonesque commented:
Sorry I didn't have time to review in detail, but apparently just using the replication rollback to rollback Time<Fixed> and Time<Virtual> was not enough? We still need to have some ad-hoc logic?

> I tested the fix using the replication process found in https://github.com/cBournhonesque/lightyear/issues/618

What were the results? There's no rollbacks anymore?

oxkitsune commented:
This noticeably improved FPS drops I was experiencing during rollback. However there are still rollbacks for me, although those might be related to other things

nick-e commented:
I updated the PR to correctly advance the elapsed time of the fixed time resource during a rollback.

We can't use `add_resource_rollback::<Time<Fixed>>::()` because we lose `Time<Fixed>`'s **current** overstep when the resource rollbacks. We don't care what its overstep was when rollbacked. Overstep is used to determine when FixedMain should tick based on real time. Rollback ticks aren't run in real time.

On top of that, the only way to progress `Time<Fixed>` with accurate overstep is by running `RunFixedMainLoop` because overstep is private. The issue with that is the `RunFixedMainLoop` schedule progresses `Time<Fixed>` using `Time<Virtual>` which relies on real time. We would have to keep track of the real time for every tick that needs to be rollbacked, add them up, and apply it to `Time<Virtual>` before running `RunFixedMainLoop`. It gets more complicated because `Time<Virtual>` has `max_delta` which caps how much time can change and some rollbacks may be farther behind than that.

As a rule of thumb, `add_resource_rollback()` should only be used for resources that will **only** be modified by systems run in the `FixedMain` schedule. This is because rollbacks only run the `FixedMain` schedule. `Time<Fixed>` is modified by `bevy_time::fixed::run_fixed_main_schedule()` which is run outside of the `FixedMain` schedule. I think it's worth mentioning so I added it into the `add_resource_rollback()` doc comments.

# Testing
I tested this PR by adding the following to `client::handle_character_actions()` with the `avian_3d_character` example:
```rust
if let Some(ref rb) = rollback {
    if let Some(tick) = rb.get_rollback_tick() {
        info!(
            "    rb-{:?}, delta: {}, elapsed: {}",
            tick,
            time.delta_seconds(),
            time.elapsed_seconds()
        );
    } else {
        info!(
            "not-rb-{:?}, delta: {}, elapsed: {}",
            tick_manager.tick(),
            time.delta_seconds(),
            time.elapsed_seconds()
        );
    }
} else {
    info!(
        "{:?}, delta: {}, elapsed: {}",
        tick_manager.tick(),
        time.delta_seconds(),
        time.elapsed_seconds()
    );
}
```
This function is only run within a system within the `FixedMain` schedule.

I set the latency to 100ms and a packet loss of 0.1 in order to incur multi-tick rollbacks. Here are some logs of the `avian_3d_character` client run with `cargo run --release -- client -c 1` without the PR:
```
2024-09-05T16:03:41.208296Z  INFO avian_3d_character::client: not-rb-Tick(257), delta: 0.015625, elapsed: 3.796875
2024-09-05T16:03:41.224970Z  INFO avian_3d_character::client: not-rb-Tick(258), delta: 0.015625, elapsed: 3.8125
2024-09-05T16:03:41.237516Z  INFO avian_3d_character::client: not-rb-Tick(259), delta: 0.015625, elapsed: 3.828125
2024-09-05T16:03:41.254084Z  INFO avian_3d_character::client: not-rb-Tick(260), delta: 0.015625, elapsed: 3.84375
2024-09-05T16:03:41.270763Z  INFO avian_3d_character::client: not-rb-Tick(261), delta: 0.015625, elapsed: 3.859375
2024-09-05T16:03:41.287490Z  INFO avian_3d_character::client: not-rb-Tick(262), delta: 0.015625, elapsed: 3.875
2024-09-05T16:03:41.299996Z  INFO avian_3d_character::client: not-rb-Tick(263), delta: 0.015625, elapsed: 3.890625
2024-09-05T16:03:41.312512Z  INFO avian_3d_character::client:     rb-Tick(257), delta: 0.004247744, elapsed: 3.9034278 <-- wrong delta; elapsed not rollbacked
2024-09-05T16:03:41.313247Z  INFO avian_3d_character::client:     rb-Tick(258), delta: 0.004247744, elapsed: 3.9034278 <-- elapsed not advancing
2024-09-05T16:03:41.313825Z  INFO avian_3d_character::client:     rb-Tick(259), delta: 0.004247744, elapsed: 3.9034278
2024-09-05T16:03:41.314345Z  INFO avian_3d_character::client:     rb-Tick(260), delta: 0.004247744, elapsed: 3.9034278
2024-09-05T16:03:41.314856Z  INFO avian_3d_character::client:     rb-Tick(261), delta: 0.004247744, elapsed: 3.9034278
2024-09-05T16:03:41.315427Z  INFO avian_3d_character::client:     rb-Tick(262), delta: 0.004247744, elapsed: 3.9034278
2024-09-05T16:03:41.315992Z  INFO avian_3d_character::client:     rb-Tick(263), delta: 0.004247744, elapsed: 3.9034278
2024-09-05T16:03:41.318197Z  INFO avian_3d_character::client: not-rb-Tick(264), delta: 0.015625, elapsed: 3.90625
```

Here are some logs with the PR:
```
2024-09-05T16:00:26.780969Z  INFO avian_3d_character::client: not-rb-Tick(129), delta: 0.015625, elapsed: 1.828125
2024-09-05T16:00:26.792864Z  INFO avian_3d_character::client: not-rb-Tick(130), delta: 0.015625, elapsed: 1.84375
2024-09-05T16:00:26.809763Z  INFO avian_3d_character::client: not-rb-Tick(131), delta: 0.015625, elapsed: 1.859375
2024-09-05T16:00:26.826218Z  INFO avian_3d_character::client: not-rb-Tick(132), delta: 0.015625, elapsed: 1.875
2024-09-05T16:00:26.838714Z  INFO avian_3d_character::client: not-rb-Tick(133), delta: 0.015625, elapsed: 1.890625
2024-09-05T16:00:26.855291Z  INFO avian_3d_character::client: not-rb-Tick(134), delta: 0.015625, elapsed: 1.90625
2024-09-05T16:00:26.871923Z  INFO avian_3d_character::client: not-rb-Tick(135), delta: 0.015625, elapsed: 1.921875
2024-09-05T16:00:26.880931Z  INFO avian_3d_character::client:     rb-Tick(129), delta: 0.015625, elapsed: 1.828125 <-- correct delta; elapsed rollbacked
2024-09-05T16:00:26.881672Z  INFO avian_3d_character::client:     rb-Tick(130), delta: 0.015625, elapsed: 1.84375  <-- elapsed advancing correctly
2024-09-05T16:00:26.882308Z  INFO avian_3d_character::client:     rb-Tick(131), delta: 0.015625, elapsed: 1.859375
2024-09-05T16:00:26.882863Z  INFO avian_3d_character::client:     rb-Tick(132), delta: 0.015625, elapsed: 1.875
2024-09-05T16:00:26.883419Z  INFO avian_3d_character::client:     rb-Tick(133), delta: 0.015625, elapsed: 1.890625
2024-09-05T16:00:26.883931Z  INFO avian_3d_character::client:     rb-Tick(134), delta: 0.015625, elapsed: 1.90625
2024-09-05T16:00:26.884464Z  INFO avian_3d_character::client:     rb-Tick(135), delta: 0.015625, elapsed: 1.921875
2024-09-05T16:00:26.888542Z  INFO avian_3d_character::client: not-rb-Tick(136), delta: 0.015625, elapsed: 1.9375
```

There are still constant rollbacks. Still looking into that.

nick-e commented:
I added a test for verifying that the time resource is accurate during a rollback. I can confirm that it fails without this PR.

cBournhonesque commented on lightyear/src/client/prediction/rollback.rs:
```diff
@@ -568,6 +569,38 @@ pub(crate) fn prepare_rollback_resource<R: Resource + Clone>(
+    let rollback_elapsed_time = current_fixed_time.elapsed() - rollback_time_offset;
+    rollback_fixed_time.advance_to(rollback_elapsed_time - rollback_fixed_time.timestep());
+    // This sets the fixed time's delta to the correct amount.
+    rollback_fixed_time.advance_by(rollback_fixed_time.timestep());
```
I think this could have a clarifying comment.
`Time<Fixed>.delta` is set to the value provided in `advance_by` (or `advance_to`), so we want to call `advance_by(rollback_fixed_time.timestep())` at the end to set the delta value that is expected.

(I had to read the `advance_by` source to check this, but it would be good to mention it here)

nick-e replied:
Updated!

cBournhonesque commented on lightyear/src/client/prediction/rollback.rs:
```diff
@@ -590,13 +623,44 @@ pub(crate) fn run_rollback(world: &mut World) {
+
+        // Set the rollback tick's generic time resource to the fixed time
+        // resource that was just advanced.
+        *world.resource_mut::<Time>() = world.resource::<Time<Fixed>>().as_generic();
```
Why do we want to also rollback the non-fixed-time Time?

nick-e replied:
`Time` is not a non-fixed time but instead a time without a "context". It could represent fixed time or some other kind of time like `Time<Virtual>`. It's my understanding that systems within the `FixedMain` should be safe to assume that the value of the generic `Time` resource it retrieves via `Res<Time>` represents the `Time<Fixed>` resource.

cBournhonesque commented on lightyear/src/shared/time_manager.rs:
```diff
@@ -56,7 +56,7 @@ pub struct TimeManager {
     pub base_relative_speed: f32,
     /// Should we speedup or slowdown the simulation to sync the ticks?
-    /// >1.0 = speedup, <1.0 = slowdown
+    /// \>1.0 = speedup, <1.0 = slowdown
```
Is this a typo?

nick-e replied:
`/// >` is a table entry just like `/// *` is a bullet point. The rust doc comment has to escape the `>` to denote that the character is meant to represent the greater-than sign and not the start of a table entry. The linter was failing due to that issue.

nick-e replied:
https://rust-lang.github.io/rust-clippy/master/index.html#/doc_lazy_continuation

cBournhonesque commented:
Thanks for the explanation on why rollbacking Time<Fixed> + running RunFixedMainLoop wouldn't work.
I also appreciate the tests and the logs!

I'm convinced by this PR, making sure that Time<Fixed> was handled correctly during rollback was something I had completely overlooked!

I left a couple comments but otherwise I think it's ready to merge. Thanks for the awesome work!

cBournhonesque commented:
@nick-e Could a reason for the avian rollbacks be that avian is using `Time<Physics>` and `Time<Substeps>` which should be rollbacked in a similar manner to `Time<Fixed>`: https://github.com/Jondolf/avian/blob/v0.2.0/src/schedule/mod.rs#L252 ?
